File: printer.ts
export class LogPrinter {
  printLogs: string[] = [];

  print(msg: string): void {
    this.printLogs.push(msg);
  }
  getPrintFunction = () => {
    return (msg: string) => this.print(msg);
  };
  getLogs(): string[] {
    return this.printLogs;
  }
  getLastLogs(): string | undefined {
    if (this.printLogs.length === 0) {
      return undefined;
    }
    return this.printLogs[this.printLogs.length - 1];
  }
}

File: go-interpreter/printer.go
package main

type LogPrinter struct {
	PrintLogs []string
}

func NewLogPrinter() *LogPrinter {
	return &LogPrinter{}
}

func (lp *LogPrinter) Print(msg string) {
	lp.PrintLogs = append(lp.PrintLogs, msg)
}

func (lp *LogPrinter) GetPrintFunction() PrintFunc {
	return func(msg string) {
		lp.Print(msg)
	}
}

func (lp *LogPrinter) GetLogs() []string {
	return lp.PrintLogs
}

func (lp *LogPrinter) GetLastLog() *string {
	if len(lp.PrintLogs) == 0 {
		return nil
	}
	s := lp.PrintLogs[len(lp.PrintLogs)-1]
	return &s
}
